src/services/refactors/convertToEs6Module.ts

```diff
@@ -0,0 +1,567 @@
+
+            const { parent } = moduleSpecifier;
+            switch (parent.kind) {
+                case SyntaxKind.ExternalModuleReference: {
+                    const importEq = (parent as ExternalModuleReference).parent;
+                    changes.replaceNode(importingFile, importEq, makeImport(importEq.name, /*namedImports*/ undefined, moduleSpecifier.text));
+                    break;
+                }
+                case SyntaxKind.CallExpression: {
+                    const call = parent as CallExpression;
+                    if (isRequireCall(call, /*checkArgumentIsStringLiteral*/ false)) {
+                        changes.replaceNode(importingFile, parent, createPropertyAccess(getSynthesizedDeepClone(call), "default"));
+                    }
+                    break;
+                }
+            }
+        }
+    }
+
+    /** @returns Whether we converted a `module.exports =` to a default export. */
+    function convertFileToEs6Module(sourceFile: SourceFile, checker: TypeChecker, changes: textChanges.ChangeTracker, target: ScriptTarget): ModuleExportsChanged {
+        const identifiers: Identifiers = { original: collectFreeIdentifiers(sourceFile), additional: createMap<true>() };
+        const exports = collectExportRenames(sourceFile, checker, identifiers);
+        convertExportsAccesses(sourceFile, exports, changes);
+        let moduleExportsChangedToDefault = false;
+        for (const statement of sourceFile.statements) {
+            const moduleExportsChanged = convertStatement(sourceFile, statement, checker, changes, identifiers, target, exports);
+            moduleExportsChangedToDefault = moduleExportsChangedToDefault || moduleExportsChanged;
+        }
+        return moduleExportsChangedToDefault;
+    }
+
+    /**
+     * Contains an entry for each renamed export.
+     * This is necessary because `exports.x = 0;` does not declare a local variable.
+     * Converting this to `export const x = 0;` would declare a local, so we must be careful to avoid shadowing.
+     * If there would be shadowing at either the declaration or at any reference to `exports.x` (now just `x`), we must conver to:
+     *     const _x = 0;
+     *     export { _x as x };
+     * This conversion also must place if the exported name is not a valid identifier, e.g. `exports.class = 0;`.
+     */
+    type ExportRenames = ReadonlyMap<string>;
+
+    function collectExportRenames(sourceFile: SourceFile, checker: TypeChecker, identifiers: Identifiers): ExportRenames {
+        const res = createMap<string>();
+        forEachExportReference(sourceFile, node => {
+            const { text, originalKeywordKind } = node.name;
+            if (!res.has(text) && (originalKeywordKind !== undefined && isNonContextualKeyword(originalKeywordKind)
+                || checker.resolveName(node.name.text, node, SymbolFlags.Value, /*includeGlobals*/ false))) {
+                res.set(text, makeUniqueName(`_${text}`, identifiers));

```

Comment:
Isn't it redundant to prepend an underscore?

Reply:
`text` is a keyword name; it doesn't appear in the identifiers list, but still needs the underscore to avoid being a keyword. And *then* if there happens to be an identifier `_class` we need to change to `__class`.